### PR TITLE
Add support for paths with alternate keys

### LINKED ISSUE #120: Add support for alternate keys
Models containing alternate keys cannot easily be converted to OpenAPI. Corresponding paths provided by a custom path provider lead to an unhandled exception in parameter creation.

### Assemblies affected

All versions of Microsoft.OpenApi.OData.Reader.dll

### Steps to reproduce

See the following modifications of the single key test. It reflects a case for an alternate key path http://host/service/Customer(AlternateId='A1245') which was added by a custom `ODataPathProvider` handed over in the `OpenAPIConvertSettings`. The same can be achieved for multi-part alternate keys.

```diff
[Theory]
[InlineData(true)]
[InlineData(false)]
public void CreateKeyParametersForSingleKeyWorks(bool prefix)
{
    // Arrange
    EdmModel model = new EdmModel();
    EdmEntityType customer = new EdmEntityType("NS", "Customer");
    customer.AddKeys(customer.AddStructuralProperty("Id", EdmPrimitiveTypeKind.String));

+    var alternateId = customer.AddStructuralProperty("AlternateId", EdmPrimitiveTypeKind.String);
+    model.AddAlternateKeyAnnotation(customer, new System.Collections.Generic.Dictionary<string, IEdmProperty> { { "AlternateId", alternateId } });

    model.AddElement(customer);
    OpenApiConvertSettings setting = new OpenApiConvertSettings
    {
        PrefixEntityTypeNameBeforeKey = prefix
    };
    ODataContext context = new ODataContext(model, setting);
    ODataKeySegment keySegment = new ODataKeySegment(customer);
    // Act
+    var parameters = context.CreateKeyParameters(keySegment, new System.Collections.Generic.Dictionary<string, string> { { "AlternateId", "alternateId" } });
-    var parameters = context.CreateKeyParameters(keySegment);
...
```

### Expected result

No exception without considering alternate keys.

### Actual result

KeyNotFoundException in OpenApiParameterGenerator.CreateKeyParameters

### Additional detail

Given the mere example `http://host/service/Employees(ID='A1245')` from [OData Version 4.01. Part 2: URL Conventions](https://docs.oasis-open.org/odata/odata/v4.01/os/part2-url-conventions/odata-v4.01-os-part2-url-conventions.html#sec_AlternateKeys), without a model it is unclear whether one should rely on the (primary) key or an alternate key. Since the model is known for the generator, not only the primary key but also the alternative keys could be taken into account when mapping the parameter names. Definitely it should work fault tolerant with any other PathProvider then the default ODataPathProvider.

### PR DESCRIPTION
**Pull request by millicentachieng**
Fixes #120 

This PR:
- Adds new paths for entities that have alternate keys
- Renames the summary and OperationId of operations to include key name
- Add a setting to enable the generation of paths with alternate keys

### CODE REVIEW

File: src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs [hunk 1]
```diff
@@ -50,7 +50,7 @@ public static string UpperFirstChar(string input)
                 return input;
             }
 
-            char first = Char.ToUpper(input[0]);
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs:53]: No need to change these to unfocus the main priority in one PR.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 2]
```diff
@@ -516,14 +523,44 @@ private void CreateRefPath(ODataPath currentPath)
...
+        /// <param name="entityType">The entityType with alternate keys</param>
+        private void CreateAlternateKeyPath(ODataPath currentPath, IEdmEntityType entityType)
+        {
+            Utils.CheckArgumentNull(currentPath, nameof(currentPath));
+            Utils.CheckArgumentNull(entityType, nameof(entityType));
+
+            IEnumerable<IDictionary<string, IEdmProperty>> alternateKeys = _model.GetAlternateKeysAnnotation(entityType);
+            if (alternateKeys.Count() > 0)
+            {
+                foreach (var keyDict in alternateKeys)
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:550]: calling .Count() and do foreach will perform double evaluation for the IEnumerable< T >. Maybe consider to call .ToList() or .ToArray() first?

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataPathProviderTests.cs [hunk 3]
```diff
@@ -617,6 +618,89 @@ public void GetPathsWithStreamPropertyAndWithEntityHasStreamWorks(bool hasStream
             }
         }
 
+        [Fact]
+        public void GetPathsWithAlternateKeyParametersWorks()
+        {
+            string alternateKeyProperty =
+@"<Property Name=""SSN"" Type=""Edm.String""/>
+    <Annotation Term=""OData.Community.Keys.V1.AlternateKeys"">
```
Comment (xuzhg) [inline test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataPathProviderTests.cs:626]: ODL supports to use OData.Core.Vocabulary alternateKeys. And that's the recommended solution.

Consider to enable this.

**Review by xuzhg — DISMISSED** [review]

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataPathProviderTests.cs (hunk 3 above)
Comment (millicentachieng) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataPathProviderTests.cs:626]: For some reason graph metadata uses `OData.Community.Keys.V1.AlternateKey`. I'm changing the test to use `Org.OData.Core.V1.AlternateKey`.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataKeySegment.cs [hunk 4]
```diff
@@ -53,13 +58,11 @@ public override string Identifier
...
             get
             {
-                IList<string> keys = new List<string>();
-                foreach (var key in EntityType.Key())
-                {
-                    keys.Add(key.Name);
-                }
+                IList<string> keys = IsAlternateKey ? 
+                    KeyMappings.Values.ToList() : 
+                    EntityType.Key().Select(x => x.Name).ToList();
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataKeySegment.cs:63]:
```suggestion
                    EntityType.Key().Select(static x => x.Name).ToList();
```

File: src/Microsoft.OpenApi.OData.Reader/Operation/EntityDeleteOperationHandler.cs [hunk 5]
```diff
@@ -36,17 +36,19 @@ protected override void Initialize(ODataContext context, ODataPath path)
...
+            string placeHolder = $"Delete entity from {EntitySet.Name} by key ({keySegment.Identifier})";
             operation.Summary = _deleteRestrictions?.Description ?? placeHolder;
             operation.Description = _deleteRestrictions?.LongDescription;
 
             // OperationId
             if (Context.Settings.EnableOperationId)
             {
                 string typeName = entityType.Name;
-                operation.OperationId = EntitySet.Name + "." + typeName + ".Delete" + Utils.UpperFirstChar(typeName);
+                string keyName = string.Join("", keySegment.Identifier.Split(',').Select(x => Utils.UpperFirstChar(x)));
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Operation/EntityDeleteOperationHandler.cs:50]:
```suggestion
                string keyName = string.Join("", keySegment.Identifier.Split(',').Select(static x => Utils.UpperFirstChar(x)));
```

File: src/Microsoft.OpenApi.OData.Reader/Operation/EntityGetOperationHandler.cs [hunk 6]
```diff
@@ -37,17 +37,19 @@ protected override void Initialize(ODataContext context, ODataPath path)
...
             operation.Summary = _readRestrictions?.ReadByKeyRestrictions?.Description ?? placeHolder;
             operation.Description = _readRestrictions?.ReadByKeyRestrictions?.LongDescription ?? Context.Model.GetDescriptionAnnotation(entityType);
 
             // OperationId
             if (Context.Settings.EnableOperationId)
-            {
+            { 
                 string typeName = entityType.Name;
-                operation.OperationId = EntitySet.Name + "." + typeName + ".Get" + Utils.UpperFirstChar(typeName);
+                string keyName = string.Join("", keySegment.Identifier.Split(',').Select(x => Utils.UpperFirstChar(x)));
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Operation/EntityGetOperationHandler.cs:51]:
```suggestion
                string keyName = string.Join("", keySegment.Identifier.Split(',').Select(static x => Utils.UpperFirstChar(x)));
```

File: src/Microsoft.OpenApi.OData.Reader/Operation/EntityUpdateOperationHandler.cs [hunk 7]
```diff
@@ -32,17 +32,19 @@ protected override void Initialize(ODataContext context, ODataPath path)
...
+            string placeHolder = $"Update entity in {EntitySet.Name} by key ({keySegment.Identifier})";
             operation.Summary = _updateRestrictions?.Description ?? placeHolder;
             operation.Description = _updateRestrictions?.LongDescription;
 
             // OperationId
             if (Context.Settings.EnableOperationId)
             {
                 string typeName = entityType.Name;
-                operation.OperationId = EntitySet.Name + "." + typeName + ".Update" + Utils.UpperFirstChar(typeName);
+                string keyName = string.Join("", keySegment.Identifier.Split(',').Select(x => Utils.UpperFirstChar(x)));
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Operation/EntityUpdateOperationHandler.cs:46]:
```suggestion
                string keyName = string.Join("", keySegment.Identifier.Split(',').Select(static x => Utils.UpperFirstChar(x)));
```

**Review by baywet — CHANGES_REQUESTED** [review]
I also believe we should introduce a setting for that new behavior

**Comment by millicentachieng** [comment]
> I also believe we should introduce a setting for that new behavior

I had a discussion with @darrelmiller regarding this, and we thought that if there are alternate keys defined in the metadata then we ought to generate paths that utilize them as well. 

Do you think the presence of these paths is going to affect any downstream tools? CC @peombwa

**Comment by baywet** [comment]
It might increase significantly the size of the SDKs, especially for sets that are fairly "close to the root" of the API paths tree since the logic might go ahead and expand all the paths we already had under the main key.
Additionally, the "custom" has been to introduce settings as a way for consumers to control the behaviour of the conversion every time we've introduced changes that add additional paths to the resulting description. This way, should somebody complain about the new paths and not want them, they have a workaround.

**Comment by baywet** [comment]
Thanks, that's what I suspected from the code but I wasn't sure whether it was a final decision or a stopgap.

**Review by baywet — APPROVED** [review]
thanks for making the changes!

**Comment by peombwa** [comment]
> > I also believe we should introduce a setting for that new behavior
> 
> I had a discussion with @darrelmiller regarding this, and we thought that if there are alternate keys defined in the metadata then we ought to generate paths that utilize them as well.
> 
> Do you think the presence of these paths is going to affect any downstream tools? CC @peombwa

As @baywet said above, we may see a significant growth in paths. However, having the setting to control how paths with alternate keys are generated is enough to warrant any concerns with downstream tools.

**Review by peombwa — APPROVED** [review]

**Pull request merged**